### PR TITLE
[Rule] EnsureExplicitNullableTypes

### PR DESCRIPTION
**Pull request by smnandre**
CF #1710

### CODE REVIEW

**Comment by smnandre** [comment]
I may need a bit of help to finalize this (not sure if i can do anything regarding CI checks)

File: docs/rules.md [hunk 1]
```diff
@@ -375,6 +376,28 @@ composer require symfony/var-dumper
...
+```rst
+function foo(string|null $bar = null)
+```
+
+##### Invalid Examples :-1:
+
+```rst
+function foo(string $bar = null)
+```
+
```
Comment (alamirault) [inline docs/rules.md:400]: This file is fully generated, depending on rule attributes (Description, InvalidExample, ValidExample)

Locally I do `php bin/doctor-rst rules > docs/rules.md` to generate it

File: src/Rule/EnsureExplicitNullableTypes.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    public function check(Lines $lines, int $number, string $filename): ViolationInterface
+    {
+        $lines->seek($number);
+        $line = $lines->current();
+
+        if (!str_contains((string) $line->clean(), ' = null')) {
+            return NullViolation::create();
+        }
```
Comment (alamirault) [inline src/Rule/EnsureExplicitNullableTypes.php:40]: To be sure, it's only for optimization right (avoid match when code style is already good) ?

File: src/Rule/EnsureExplicitNullableTypes.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+        if ($matches = $line->clean()->match($pattern)) {
+            $types = $matches[1];
+
+            // ?int $id = null
+            if (str_starts_with($types, '?')) {
+                return NullViolation::create();
+            }
+
+            // int|null $id = null
```
Comment (alamirault) [inline src/Rule/EnsureExplicitNullableTypes.php:52]: Do we accept this syntax in symfony-docs ?

If not, we could remove it from rule, or make rule configurable (see `src/Rule/DeprecatedDirectiveMajorVersion.php` wichi implements `Configurable`)

@OskarStark What's your opinion?

File: tests/Rule/EnsureExplicitNullableTypesTest.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /**
+     * @return \Generator<string, array{0: ViolationInterface, 1: RstSample}>
+     */
+    public static function validProvider(): iterable
+    {
+        $validCases = [
+            'function foo(?int $bar = null)',
+            'function foo(int|null $bar = null)',
```
Comment (alamirault) [inline tests/Rule/EnsureExplicitNullableTypesTest.php:46]: We should add cases with multiple parameters, and according to the regex, multiple or no spaces

`'function foo(int|null $bar = null, int $baz = null)'` is considered valid but is not

We can also add `'function foo($payload = null)'` as valid

**Review by alamirault — CHANGES_REQUESTED** [review]
Thank you @smnandre for this new rule !

I've made some suggestions

File: src/Rule/EnsureExplicitNullableTypes.php (hunk 3 above)
Comment (OskarStark) [reply src/Rule/EnsureExplicitNullableTypes.php:52]: In Symfony we use ?int but also int|string|null 

Let's keep it for now

File: tests/Rule/EnsureExplicitNullableTypesTest.php (hunk 4 above)
Comment (OskarStark) [reply tests/Rule/EnsureExplicitNullableTypesTest.php:46]: I would only handle Parma's with types

**Comment by OskarStark** [comment]
Right now it runs on every RST file line. We should check in the rule if we are in a Php code block.
Please check the inspiration PR

File: docs/rules.md (hunk 1 above)
Comment (smnandre) [reply docs/rules.md:400]: Oh thanks! (and great DX btw 👏 )

File: src/Rule/EnsureExplicitNullableTypes.php (hunk 2 above)
Comment (smnandre) [reply src/Rule/EnsureExplicitNullableTypes.php:40]: Yep

Comment (smnandre) [reply src/Rule/EnsureExplicitNullableTypes.php:40]: Well "not concerned" more than "good" but yes... Do you find it superfluous?

File: src/Rule/EnsureExplicitNullableTypes.php (hunk 3 above)
Comment (smnandre) [reply src/Rule/EnsureExplicitNullableTypes.php:52]: Probably not so frequent for just one type, but it's more to illustrate the syntax when multiple types + null are set. 

```php
final class MonotonicClock implements ClockInterface
{
    // ...

    public function __construct(\DateTimeZone|string|null $timezone = null)
    {
```

I'm replacing the comment with "int|string|null" to be more accurate

File: src/Rule/EnsureExplicitNullableTypes.php (hunk 2 above)
Comment (OskarStark) [reply src/Rule/EnsureExplicitNullableTypes.php:40]: No it's fine

File: tests/Rule/EnsureExplicitNullableTypesTest.php (hunk 4 above)
Comment (smnandre) [reply tests/Rule/EnsureExplicitNullableTypesTest.php:46]: ```function foo($payload = null)``` is something we maybe don't want to "promote", what do you think ? 

In all manner, it won't raise violation here

---

`function foo(int|null $bar = null, int $baz = null)`

Thanks i'll update the code (maybe could we add a matchAll() method there ?)

Comment (smnandre) [reply tests/Rule/EnsureExplicitNullableTypesTest.php:46]: Oh nevermind, i've read the code 😮‍💨

**Comment by smnandre** [comment]
> Right now it runs on every RST file line. We should check in the rule if we are in a Php code block. Please check the inspiration PR

Thanks, just needed to understand the lines / number / block logic :)

**Review by OskarStark — APPROVED** [review]

**Pull request merged**